Add a named_call public function in lax

qiuminxu opened the pull request:
Add a named_call public function in lax.

The implementation copies from the flax [_named_call](https://github.com/google/flax/blob/bb3170165010d508e7950d2c66eb9846d3661471/flax/core/lift.py#L674)

[Tests](https://github.com/deepmind/dm-haiku/blob/3fedc4725da8f58b4dae80bb21b195b0e3d9b670/haiku/_src/named_call_test.py#L25) passed after replacing the haiku [named_fun](https://github.com/deepmind/dm-haiku/blob/3fedc4725da8f58b4dae80bb21b195b0e3d9b670/haiku/_src/named_call.py#L109) with this implementation.

qiuminxu committed:
Add named_call public API.

Move named_call_p to core.py from lax.py.
Also move the translation rule to jax/interpreters/xla.py where the core_call translation rule is.

qiuminxu commented:
cc @froystig, @LenaMartens @tomhennigan @trevorcai @levskaya @jheek @avital

Thanks @LenaMartens for verifying this implementation also has raised the abstraction level of the arguments. Looks like the two implementation are equivalent. Are we good to upstream this implementation or would you prefer to copy the implementation to double check and then upstream?

tomhennigan commented:
If @LenaMartens is happy I'm happy 😄

tomhennigan commented on jax/_src/lax/lax.py:
```diff
@@ -2468,6 +2468,16 @@ def _named_call_translation_rule(c, axis_env, in_nodes, name_stack,
 xla.call_translations[named_call_p] = _named_call_translation_rule
 
+
+def named_call(f, name):
```
Doc string? This is a new public symbol after all 😄 

We should also be adding this to https://github.com/google/jax/blob/master/docs/jax.lax.rst.

qiuminxu replied:
Done.

tomhennigan reviewed (changes requested):
Looks good to me, but documentation would be very welcome.

tomhennigan commented on docs/jax.lax.rst:
```diff
@@ -162,6 +162,14 @@ Custom gradient operators
     custom_linear_solve
     custom_root
 
+Function operators
```
Is `named_call` an operator (I understand that the symbol is a function that operates on functions, but since it is a higher order function I find the word "operator" a bit confusing, might just be me..)?

If we keep this in `jax.lax` would put this at the bottom of the file under a heading `Debugging` or `Compilation`. Alternatively we could move this to `jax.named_call` and then document it under `Just-in-time-compilation (jit)`. I think that might actually make the most sense since `named_call` is only useful once JAX programs are staged out.

@gnecula wdyt?

gnecula replied:
I like the idea of putting this in `jax` instead of `lax`.

qiuminxu replied:
Done. I created a `named_call.py` under `jax/_src` which has the implementation, and another `jax/named_call.py` that only contains only imports. Then I exposed both symbols directly under jax (`jax.named_call` and `jax.named_call_p` )

tomhennigan commented on jax/_src/lax/lax.py:
```diff
@@ -2468,6 +2468,33 @@ def _named_call_translation_rule(c, axis_env, in_nodes, name_stack,
+    *,
+    name: Optional[str] = None,
+) -> Callable[..., Any]:
+  """Wraps a function in a name_scope with the provided name.
```
JAX does not have the concept of a "name scope" and we should not assume users are familiar with TF before using JAX. I think the important thing to call out is that named call is used to add names to staged jax programs.

Here's a concrete suggestion of a perhaps more useful doc string:

```
Adds a user specified name to a function when staging out JAX computations.

When staging out computations for just-in-time compilation to XLA (or other
backends such as TensorFlow) JAX runs your Python program but by default does
not preserve any of the function names or other metadata associated with it.
This can make debugging the staged out (and/or compiled) representation of your
program complicated because there is limited context information for each
operation being executed.

`named_call` tells JAX to stage the given function out as a subcomputation with
a specific name. When the staged out program is compiled with XLA these named
subcomputations are preserved and show up in debugging utilities like the
TensorFlow Profiler in TensorBoard. Names are also preserved when staging out
JAX programs to TensorFlow using :func:`experimental.jax2tf.convert`.
```

gnecula replied:
I like this. Ideally we should have a documentation of how to interpret the name_stack that shows up in xprof.

qiuminxu replied:
Thanks for the doc string, it looks very nice! 

> I like this. Ideally we should have a documentation of how to interpret the name_stack that shows up in xprof.

Should we add a screenshot and some explanations in this page? https://jax.readthedocs.io/en/latest/profiling.html?

tomhennigan commented on jax/_src/lax/lax.py:
```diff
@@ -2468,6 +2468,33 @@ def _named_call_translation_rule(c, axis_env, in_nodes, name_stack,
+    name = fun.__name__
+
+  _, in_tree = jax.tree_flatten(())
+  def named_f(*args, **kwargs):
```
nit: Should we `@functools.wraps(fun)` to preserve docstring et al?

qiuminxu replied:
Done.

tomhennigan reviewed (changes requested):
Looks good, final comments around documentation and location.

gnecula commented:
I am not sure it is a good idea to have named_call in its own file. It is a rather small feature to deserve a file on its own. Perhaps it should go into `api`, even though it is kind of large? @hawkinsp has reorganized the files recently, at least partially. I am wondering what he thinks is the right place for this going forward. 

Since you added `named_call_test.py` this PR will need some changes to the tests/BUILD file. If, however, we merge named_call somewhere else, then it makes sense to merge the tests also.

froystig commented:
Let's add the implementation (definitions of primitive and rules) to `core.py`, right below the corresponding implementation of `call_p`. We can then import `named_call` in `api.py` so that it is available externally as `jax.named_call`.

qiuminxu commented:
> I am not sure it is a good idea to have named_call in its own file. It is a rather small feature to deserve a file on its own. Perhaps it should go into `api`, even though it is kind of large? @hawkinsp has reorganized the files recently, at least partially. I am wondering what he thinks is the right place for this going forward.
> 
> Since you added `named_call_test.py` this PR will need some changes to the tests/BUILD file. If, however, we merge named_call somewhere else, then it makes sense to merge the tests also.

Done. I moved named_call_p close to call_p. Also added named_call public function in api.py. Tests are also added to api_test.py.

qiuminxu committed:
Update core.py

qiuminxu commented:
@tomhennigan @gnecula @froystig Does the latest version look good?

gnecula reviewed (approved):
Thanks!
